docs(spirekey): add documentation for connecting a dapp to Kadena SpireKey (#1706)
kadena-community/kadena.js: packages/apps/docs/src/docs/kadena/wallets/kadena-spirekey/index.md

---

> title: Kadena SpireKey User Guide

Author: I changed it to Authentication and Signing

description: Integrate Kadena SpireKey in your dApp
menu: Kadena SpireKey
label: Kadena SpireKey User Guide
order: 2
layout: full
---

# Kadena SpireKey User Guide

---

- [Kadena SpireKey User Guide](#kadena-spirekey-user-guide)
  - [Short Introduction](#short-introduction)
  - [Conceptual Overview](#conceptual-overview)
    - [Create account in Kadena SpireKey](#create-account-in-kadena-spirekey)
    - [Login with Kadena SpireKey](#login-with-kadena-spirekey)
    - [Sign Transaction from dApp](#sign-transaction-from-dapp)
  - [Specification](#specification)
    - [OAuth authenticate](#oauth-authenticate)
      - [Authentication request](#authentication-request)
      - [Authentication Response](#authentication-response)
    - [Kadena SpireKey Sign](#kadena-spirekey-sign)
      - [Sign Request](#sign-request)
      - [Sign Response](#sign-response)

## Short Introduction

> 
> In this document you will learn how to **connect** your dApp to Kadena SpireKey,
> **retrieve the available accounts** and **sign** with SpireKey Connect
> 
> ## Conceptual Overview
> 
> From a dApp we want to access the accounts that are available on the Kadena
> SpireKey Wallet. With this information, we can then construct transactions.
> These transactions need to be signed by Kadena SpireKey
> 
> We're all familiar with OAuth. OAuth enables the user to authenticate with a
> third-party authentication provider like Google, Facebook, Github and many more.
> 
> ![OAuth Flow](image.png)
> 
> The similarities are:
> 
> - the **dApp** is the **Client Application**
> - **Kadena SpireKey** works as a delegate for **Authorization Server** and in
>   turn acesses **WebAuthN** to grant access
> - and the **dApp** is hosted by the **Resource Server**
> 
> Let's take a look in how the flow would be for a **dApp** and **Kadena
> Spirekey**
> 

Reviewer: ```suggestion

```

I'm finding this hard to follow. See my somewhat mangled work-in-progress replacement text.

> 1. The user creates an account on Kadena SpireKey by authenticating with
>    WebAuthN
> 2. The user login with Kadena SpireKey from the dApp
> 3. The user wants to sign a transaction created by the dApp
> 

Reviewer: ```suggestion
## Integrate with Kadena SpireKey

There are three basic integration points between a decentralized application and Kadena SpireKey.
To integrate with Kadena SpireKey, your application needs to enable users to perform the following steps:

 - Register at least one account on Kadena SpireKey by authenticating using a Web Authentication (WebAuthN) method.
- Log in using a Kadena SpireKey account with authentication provided by a Web Authentication (WebAuthN) method.
- Sign transactions created by the application using a Kadena SpireKey account with authentication provided by a Web Authentication (WebAuthN) method.
```

> ### Create account in Kadena SpireKey

Reviewer: ```suggestion
### Register an account in Kadena SpireKey
```

> 

Reviewer: ```suggestion
The following diagram illustrates the process for registering an account using Kadena SpireKey.
```

```mermaid
sequenceDiagram
  participant u as User
  participant s as Kadena SpireKey
  participant w as WebAuthN
  participant k as Kadena Blockchain
    u->>s: create account
    s->>w: request authentication token
    w->>s: `WebAuthN-token`
    s->>k: create account with `WebAuthN-token`
  alt Validation by blockchain
    k->>k: executing transaction, mining block
    k->>s: account created `c:my-account`
  else Validation failed
    k->>s: transaction invalid
  end
```

> ### Login with Kadena SpireKey
> 

Reviewer: ```suggestion
### Log in with Kadena SpireKey

The following diagram illustrates the process for logging in with a Kadena SpireKey account.
 ```

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: Login with SpireKey
  create participant s as Kadena SpireKey
    d->>s: navigate spirekey.kadena.io/login?returnUrl=dapp.com/auth
    u->>s: grant authorization `base-64-user`
    s->>d: dapp.com/auth?user=base-64-user
    d->>u: logged in with `base-64-user`
```

> ### Sign Transaction from dApp
> 

Reviewer: ```suggestion
### Sign a transaction from the application

The following diagram illustrates the process for signing a transaction using a Kadena SpireKey account.
```

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: buy cookie
    d->>d: create buy-cookie-transaction `tx-base64`
  create participant s as Kadena SpireKey
    d->>s: spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=dapp.com/auth
    u->>s: grant signature `tx-base64-signature`
    s->>d: dapp.com/auth?transaction=tx-base64-signature
  create participant k as Kadena Blockchain
    d->>k: send transaction
```

## Specification

### Kadena SpireKey authenticate/login

#### Authentication request

A GET request needs to be send to Kadena SpireKey

```
spirekey.kadena.io/login?returnUrl=your-dapp.com/auth

> ```

Reviewer: ```suggestion

## Authentication request and response

To authenticate user requests for access to your application, your application must send a GET request to the Kadena SpireKey endpoint and receive a response.

### Authentication request

Send a GET request to the Kadena SpireKey `login` endpoint using the following format:

https://spirekey.kadena.io/login?returnUrl=your-dapp.com/auth

- host `spirekey.kadena.io`
- path `/login`
- querystring parameters
  - `returnUrl` the url where the user will be redirected to after
    authentication.

> #### Authentication Response
> 

Reviewer: ```suggestion
### Authentication response

Your application receives a response with the following payload:

```

```
your-dapp.com/auth?user=base-64-user
```

- `user`: `base64<JSON>` - base-64 JSON object, from `user` querystring
  parameter
  - `credentials`: `Array<object>`
    - `publicKey`: `string` - The public-key associated with the account
    - `type`: `string` - type of the credential, can be `WebAuthN`
    - `id`: `string` (optional) - The WebAuthN Credential ID. Omitted when
      `type=ED25519`
  - `accountName`: `string` - the c-account of the user
  - `name`: `string` - Alias the user provided when creating the account in
    Kadena SpireKey
  - `pendingTxIds`: `Array<string>` - List of transaction Request Keys. One of:
    - Account Creation Transaction id: the account is being minted, and the dapp
      can use this to verify when this is completed.  
      The `user` object, can already be used to prepare transactions for the
      user.

> ### Kadena SpireKey Sign
> 
> #### Sign Request
> 

Reviewer: ```suggestion
## Sign request and response

To sign transactions for your application, your application must send the transaction to the Kadena SpireKey endpoint and receive a response.

### Sign request

Send a request to the Kadena SpireKey `sign` endpoint using the following format:

https://spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=your-dapp.com/auth
```

> ```
> spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=your-dapp.com/auth
> ```

Reviewer: ```suggestion
```

- host `spirekey.kadena.io`
- path `/sign`
- querystring parameters
  - `transaction`: [`IUnsignedCommand`]() a base64 encoded transaction
  - `returnUrl` the url where the user will be redirected to after signing
  - `optimistic`: `boolean` - when `true`, Kadena SpireKey will return a signed
    transaction, even when there are pending transactions.  
    The response will include `pendingTxIds`.

> #### Sign Response
> 

Reviewer: ```suggestion
### Sign response

Your application receives a response with the following payload:

```

  else Validation failed
    k->>s: transaction invalid
  end
```

### Login with Kadena SpireKey

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: Login with SpireKey
  create participant s as Kadena SpireKey
    d->>s: navigate spirekey.kadena.io/login?returnUrl=dapp.com/auth
    u->>s: grant authorization `base-64-user`
    s->>d: dapp.com/auth?user=base-64-user
    d->>u: logged in with `base-64-user`
```

### Sign Transaction from dApp

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: buy cookie
    d->>d: create buy-cookie-transaction `tx-base64`
  create participant s as Kadena SpireKey
    d->>s: spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=dapp.com/auth
    u->>s: grant signature `tx-base64-signature`
    s->>d: dapp.com/auth?transaction=tx-base64-signature
  create participant k as Kadena Blockchain
    d->>k: send transaction
```

## Specification

### OAuth authenticate

#### Authentication request

A GET request needs to be send to Kadena SpireKey

```
spirekey.kadena.io/login?returnUrl=dapp.com/auth
```

- host `spirekey.kadena.io`
- path `/login`
- querystring parameters

>   - `returnUrl` the url where the user will be redirected to after

Reviewer 2: Maybe just a note for the developers that it should include https protocol

    k->>s: account created `c:my-account`
  else Validation failed
    k->>s: transaction invalid
  end
```

### Log in with Kadena SpireKey

The following diagram illustrates the process for logging in with a Kadena SpireKey account.
```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: Login with SpireKey
  create participant s as Kadena SpireKey
    d->>s: navigate spirekey.kadena.io/login?returnUrl=dapp.com/auth
    u->>s: grant authorization `base-64-user`
    s->>d: dapp.com/auth?user=base-64-user
    d->>u: logged in with `base-64-user`
```

### Sign a transaction from the application

The following diagram illustrates the process for signing a transaction using a Kadena SpireKey account.
```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: buy cookie
    d->>d: create buy-cookie-transaction `tx-base64`
  create participant s as Kadena SpireKey
    d->>s: spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=dapp.com/auth
    u->>s: grant signature `tx-base64-signature`
    s->>d: dapp.com/auth?transaction=tx-base64-signature
  create participant k as Kadena Blockchain
    d->>k: send transaction
```

> 

Author: I'm writing a minimal implementation as a showcase. Maybe we can take parts from there: https://github.com/kadena-community/kadena.js/pull/1768/files#diff-899d8c6b610edd0a215fe8b8190afe883bfea0c1173a5fe135540041fc05b875

Let's take a look in how the flow would be for a **dApp** and **Kadena
Spirekey**

## Integrate with Kadena SpireKey

There are three basic integration points between a decentralized application and
Kadena SpireKey. To integrate with Kadena SpireKey, your application needs to
enable users to perform the following steps:

- Register an account on Kadena SpireKey by authenticating in Kadena SpireKey
  using a Web Authentication (WebAuthn) method.
- Log in using a Kadena SpireKey account.
- Sign transactions created by the application using a Kadena SpireKey account.

### Register an account in Kadena SpireKey

The following diagram illustrates the process for registering an account using
Kadena SpireKey.

```mermaid
sequenceDiagram
  participant u as User
  participant s as Kadena SpireKey
  participant w as WebAuthn
  participant k as Kadena Blockchain
    u->>s: create account
    s->>w: request authentication token
    w->>s: `WebAuthn-token`
    s->>k: create account with `WebAuthn-token`
  alt Validation by blockchain
    k->>k: executing transaction, mining block
    k->>s: account created `c:my-account`
  else Validation failed
    k->>s: transaction invalid
  end
```

> 

Reviewer: Need some descriptive text to explain what the diagram is trying to convey.

### Log in with Kadena SpireKey

The following diagram illustrates the process for logging in with a Kadena
SpireKey account.

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: Login with SpireKey
  create participant s as Kadena SpireKey
    d->>s: navigate https://spirekey.kadena.io/login?returnUrl=https://your-dapp.com/auth
    s-->>u: request authorization
    u->>s: grant authorization `base-64-user`
    s->>d: https://your-dapp.com/auth?user=base-64-user
    d->>u: logged in with `base-64-user`
```

> 

Reviewer: Need some descriptive text to explain what the diagram is trying to convey.

### Sign a transaction from the application

The following diagram illustrates the process for signing a transaction using a
Kadena SpireKey account.

```mermaid
sequenceDiagram
  participant u as User
  participant d as dApp
    u->>d: buy cookie
    d->>d: create buy-cookie-transaction `tx-base64`
  create participant s as Kadena SpireKey
    d->>s: https://spirekey.kadena.io/sign?transaction=tx-base64&returnUrl=https://your-dapp.com/auth
    u->>s: grant signature `tx-base64-signature`
    s->>d: https://your-dapp.com/auth?transaction=tx-base64-signature
  create participant k as Kadena Blockchain
    d->>k: send transaction
```

> ## Specification
> 
> ### OAuth authenticate
> 
> #### Authentication request
> 
> A GET request needs to be send to Kadena SpireKey
> 

Reviewer: ```suggestion
## Authentication request and response

To authenticate users for your application, your application must send a GET request to the Kadena SpireKey `login` endpoint and receive a response.

### Authentication request

Send a GET request to Kadena SpireKey using the HTTPS protocol.

```
